clients/src/test/java/org/apache/kafka/clients/NetworkClientApiVersionsCheckTest.java

```diff
@@ -0,0 +1,108 @@
+
+    private static final List<Short> USED_API_KEYS = Arrays.asList(
+            ApiKeys.METADATA.id,
+            ApiKeys.FETCH.id,
+            ApiKeys.GROUP_COORDINATOR.id,
+            ApiKeys.HEARTBEAT.id,
+            ApiKeys.JOIN_GROUP.id,
+            ApiKeys.LEAVE_GROUP.id,
+            ApiKeys.LIST_OFFSETS.id,
+            ApiKeys.OFFSET_COMMIT.id,
+            ApiKeys.OFFSET_FETCH.id,
+            ApiKeys.SYNC_GROUP.id);
+
+    @Override
+    protected Collection<ApiVersionsResponse.ApiVersion> getExpectedApiVersions() {
+        List<ApiVersionsResponse.ApiVersion> expectedApiVersions = new ArrayList<>();
+        for (Short apiKey: this.USED_API_KEYS)
+            expectedApiVersions.add(new ApiVersionsResponse.ApiVersion(apiKey, Protocol.MIN_VERSIONS[apiKey], Protocol.CURR_VERSION[apiKey]));
+        return expectedApiVersions;
+    }
+
+    @Test
+    public void testUnsupportedLesserApiVersions() {
+        unsupportedApiVersionsCheck(getExpectedApiVersions(), (short) (Short.MAX_VALUE - 2), Short.MAX_VALUE);
+    }
+
+    @Test
+    public void testUnsupportedGreaterApiVersions() {
+        unsupportedApiVersionsCheck(getExpectedApiVersions(), Short.MIN_VALUE, (short) (Short.MAX_VALUE + 2));
+    }
+
+    @Test
+    public void testUnsupportedMissingApiVersions() {
+        unsupportedApiVersionsCheck(new ArrayList<ApiVersionsResponse.ApiVersion>(), Short.MIN_VALUE, (short) (Short.MAX_VALUE + 2));
+    }
+
+    private void unsupportedApiVersionsCheck(final Collection<ApiVersionsResponse.ApiVersion> expectedApiVersions, short minVersion, short maxVersion) {
+        ResponseHeader respHeader = new ResponseHeader(0);
+        Struct resp = new Struct(ProtoUtils.currentResponseSchema(ApiKeys.API_VERSIONS.id));
+        resp.set(ERROR_CODE_KEY_NAME, (short) 0);
+        List<Struct> apiVersionList = new ArrayList<>();
+        for (ApiVersionsResponse.ApiVersion apiVersion : expectedApiVersions) {
+            Struct apiVersionStruct = resp.instance(API_VERSIONS_KEY_NAME);
+            apiVersionStruct.set(API_KEY_NAME, apiVersion.apiKey);
+            apiVersionStruct.set(MIN_VERSION_KEY_NAME, minVersion);
+            apiVersionStruct.set(MAX_VERSION_KEY_NAME, maxVersion);
+            apiVersionList.add(apiVersionStruct);
+        }
+        resp.set(API_VERSIONS_KEY_NAME, apiVersionList.toArray());
+        int size = respHeader.sizeOf() + resp.sizeOf();
+        ByteBuffer buffer = ByteBuffer.allocate(size);
+        respHeader.writeTo(buffer);
+        resp.writeTo(buffer);
+        buffer.flip();
+        selector.delayedReceive(new DelayedReceive(node.idString(), new NetworkReceive(node.idString(), buffer)));
+        try {
+            while (!client.ready(node, time.milliseconds()))
+                client.poll(1, time.milliseconds());
+            fail("Api versions check failed.");
+        } catch (KafkaException kex) {
+            // Expected
```

Comment:
Not sure, if you wanted me to add checks for exception messages, as it makes these tests a little brittle. I anyway, went ahead and added those checks. Let me know if you think they don't add much.